core/primitives/src/stateless_validation.rs

```diff
@@ -24,16 +25,31 @@ type SignatureDifferentiator = String;
 #[derive(Debug, Clone, PartialEq, Eq, BorshSerialize, BorshDeserialize)]
 pub struct EncodedChunkStateWitness(Box<[u8]>);
 
+pub type ChunkStateWitnessSize = usize;
+
 impl EncodedChunkStateWitness {
-    pub fn encode(witness: &ChunkStateWitness) -> Self {
-        Self(borsh::to_vec(witness).expect("borsh serialization should not fail").into())
+    /// Borsh-serialize and compress state witness.
+    /// Returns encoded witness along with the raw (uncompressed) witness size.
+    pub fn encode(witness: &ChunkStateWitness) -> std::io::Result<(Self, ChunkStateWitnessSize)> {
+        const STATE_WITNESS_COMPRESSION_LEVEL: i32 = 3;
+        let borsh_bytes = borsh::to_vec(witness)?;
```

Comment:
How large are these `witness`es? If they are of appreciable size, I would recommend streaming this data instead (https://docs.rs/borsh/latest/borsh/fn.to_writer.html and https://docs.rs/zstd/latest/zstd/stream/write/struct.Encoder.html seem relevant.)

Reply:
It can be as large as max network message size which is 500MB :(
Streaming is a great idea, I've tried implementing it and counting written bytes appeared to somewhat non-trivial.
I think it is not critical for MVP, so I'm merging it as it is for now as agreed [here](https://near.zulipchat.com/#narrow/stream/407237-core.2Fstateless-validation/topic/Faster.20decision.20making/near/432338003). Also I've created https://github.com/near/nearcore/issues/11064 to address this in the future.